src/align/models.jl

```diff
@@ -0,0 +1,97 @@
+# Score Models
+# ------------
+
+"""
+Supertype of score model.
+
+Every score model is a problem of finding the maximum score and its alignment.
+"""
+abstract AbstractScoreModel{T<:Real}
+
+"""
+Affine gap scoring model.
+
+The gap penalty of length `k` is `gap_open_penalty + gap_extend_penalty * k`.
+
+Fields:
+
+    * `submat`: a substitution matrix
+    * `gap_open_penalty`: a penalty of opening a new gap
+    * `gap_extend_penalty`: a penalty of extending a gap
+"""
+type AffineGapScoreModel{T} <: AbstractScoreModel{T}
+    submat::AbstractSubstitutionMatrix{T}
+    gap_open_penalty::T
+    gap_extend_penalty::T
+end
+
+function AffineGapScoreModel{T}(submat::AbstractSubstitutionMatrix{T},
+                                gap_open_penalty, gap_extend_penalty)
+    return AffineGapScoreModel{T}(submat, T(gap_open_penalty), T(gap_extend_penalty))
+end
+
+function AffineGapScoreModel{T}(submat::AbstractSubstitutionMatrix{T};
+                                gap_open_penalty=nothing, gap_extend_penalty=nothing)
+    if gap_open_penalty === nothing || gap_extend_penalty === nothing
```

Comment:
Yes, but I just want the keyword arguments because users (including me) cannot remember the order of arguments. This interface would be controversial and I'll offer some candidates later.